### ISSUE TITLE
Unable to cancel pending request when using RxJS pipe

### ISSUE THREAD

**Issue by tim-kuteev** [angular/angular#27411]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/common/http

### Is this a regression?

I was able to cancel request using old way chained rxjs operators.

### Description

I'm not able to cancel pending request when I use **.pipe** in between. Either I try to **unsubscribe** or use 
**takeUntil** operator, the request remains pending. You can see it on the Network tab of chromes DevTools and on the server side (if you upload multipart for example, you can see that the progress continues).

## 🔬 Minimal Reproduction

I'm not able to reproduce it on stackblitz since it uses some kind of wrap for the requests. I created repo with the example:
https://github.com/tim-kuteev/ng7-pending-request-issue

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular CLI: 7.0.7
Node: 8.12.0
OS: win32 x64
Angular: 7.1.1
... common, compiler, core, forms, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.10.7
@angular-devkit/build-angular     0.10.7
@angular-devkit/build-optimizer   0.10.7
@angular-devkit/build-webpack     0.10.7
@angular-devkit/core              7.0.7
@angular-devkit/schematics        7.0.7
@angular/cli                      7.0.7
@angular/compiler-cli             7.0.4
@angular/language-service         7.0.4
@ngtools/webpack                  7.0.7
@schematics/angular               7.0.7
@schematics/update                0.10.7
rxjs                              6.3.3
typescript                        3.1.6
webpack                           4.19.1</code></pre>

<pre><code>Google Chrome Version 69.0.3497.100 (Official Build) (64-bit)</code></pre>

**Comment by ChrisDahmCS** [comment]
I don't think this is an angular issue, I think it's an RXJS issue and how pipe handles subscriptions for you. Currently with pipe, there isn't a way that I know of to cancel early. The problem is, when it emits the value through the pipe is when it'll hit the operators, so having the takeUntil will only matter once the subscription internal to the pipe emits another value.

Basically, pipe = fancy subscribe
so request
=> subscribe
=> emit value
=> operators

Ending a pipe early could be a really neat feature request for RXJS- as I think many people have this same issue :).

**Comment by tim-kuteev** [comment]
Hi **ChrisDahmCS**

Thank you for your reply, but I don't think this is entirely correct:

> The problem is, when it emits the value through the pipe is when it'll hit the operators, so having the takeUntil will only matter once the subscription internal to the pipe emits another value.

Following this logic the following code would hit **tap** twice before knowing that takeUntil was notified.

```JavaScript
const rxjs = require('rxjs');
const operators = require('rxjs/operators');
const complete = new rxjs.Subject();
debug('start');

rxjs.interval(100)
  .pipe(
    operators.tap(val => console.log('tap %o', val)),
    operators.takeUntil(complete)
  )
  .subscribe();

rxjs.of(null)
  .pipe(operators.delay(150))
  .subscribe(() => {
    console.log('complete');
    complete.next();
  });
```
You can even multiply **pipe**s to make sure that they are chained and aware of subscription's complete.

But lets step away from **takeUntil**, the common **unsubscribe** is the core functionality of RxJS. I assume it should notify the emitter about completion, no matter using pipes or not. I don't think RxJS team just gave up on that when they moved from chains to pipes, it would be a big memory leak flaw.

I still consider that RxJS works as expected here and it's **HttpClient**'s responsibility to abort the request when the subscription gets complete.

**Comment by ChrisDahmCS** [comment]
Hey @tim-kuteev,

Thanks for the detailed response!

First I'd like to apologize to RXJS for suggesting there was a bug in somewhere in there. But on the same token, I am still going to say this is not a bug in Angular's HttpClient either-

While doing [this reproduction](https://stackblitz.com/edit/angular-fixed-imports), I noticed the rxjs imports were a bit strange. Also with StackBlitz, there are two requests made like you mentioned, but you can ignore the one that is the fetch request from the workbox-core.prod.js.

After changing `import { Subject, Subscription } from 'rxjs/index';` to `import { Subject, Subscription } from 'rxjs';` and `import { takeUntil, tap } from 'rxjs/internal/operators';` to `import { takeUntil, tap } from 'rxjs/operators';`, everything works as expected.

Do you get the same results?

Cheers!
